sdks/java/extensions/join-library/src/main/java/org/apache/beam/sdk/extensions/joinlibrary/Join.java

```diff
@@ -350,6 +368,255 @@ public void processElement(ProcessContext c) {
+      this.temporalBound = temporalBound;
+      this.compareFn = compareFn;
+      this.evictionFrequency =
+          temporalBound.getMillis() <= 4 ? Duration.millis(1) : temporalBound.dividedBy(4);
+    }
+
+    @ProcessElement
+    public void processElement(
+        ProcessContext c,
+        @AlwaysFetched @StateId("left") OrderedListState<V1> leftState,
+        @AlwaysFetched @StateId("right") OrderedListState<V2> rightState,
+        @AlwaysFetched @StateId("lastEviction") ValueState<Instant> lastEvictionState,
+        @Timestamp Instant timestamp,
+        @TimerId("eviction") Timer evictionTimer) {
+      Instant lastEviction = lastEvictionState.read();
+      if (lastEviction == null) {
+        // Initialize timer for the first time relatively since event time watermark is unknown.
+        evictionTimerSet = true;
+        evictionTimer.offset(evictionFrequency).setRelative();
+      } else if (!evictionTimerSet) {
+        // Set timer using persisted event watermark from last timer firing event time.
+        checkNotNull(lastEviction);
+        evictionTimerSet = true;
+        evictionTimer.set(lastEviction.plus(evictionFrequency));
+      }
+
+      KV<K, KV<V1, V2>> e = c.element();
+      K key = e.getKey();
+      V1 left = e.getValue().getKey();
+      V2 right = e.getValue().getValue();
+      if (left != null) {
+        leftState.add(TimestampedValue.of(left, timestamp));
+        rightState
+            .readRange(timestamp.minus(temporalBound), timestamp.plus(temporalBound))
+            .forEach(
+                r -> {
+                  KV<V1, V2> matchCandidate = KV.of(left, r.getValue());
+                  if (new Duration(r.getTimestamp(), timestamp).abs().isShorterThan(temporalBound)
+                      && compareFn.apply(matchCandidate)) {
+                    c.output(KV.of(key, matchCandidate));
+                  }
+                });
+      } else {
+        rightState.add(TimestampedValue.of(right, timestamp));
+        leftState
+            .readRange(timestamp.minus(temporalBound), timestamp.plus(temporalBound))
+            .forEach(
+                l -> {
+                  KV<V1, V2> matchCandidate = KV.of(l.getValue(), right);
+                  if (new Duration(l.getTimestamp(), timestamp).abs().isShorterThan(temporalBound)
+                      && compareFn.apply(matchCandidate)) {
+                    c.output(KV.of(key, matchCandidate));
+                  }
+                });
+      }
+    }
+
+    @OnTimer("eviction")
+    public void onEviction(
+        @StateId("left") OrderedListState<V1> leftState,
+        @StateId("right") OrderedListState<V2> rightState,
+        @StateId("lastEviction") ValueState<Instant> lastEvictionState,
+        @Timestamp Instant ts) {
+      evictionTimerSet = false;
+      lastEvictionState.write(ts);
+      leftState.clearRange(new Instant(0L), ts);
+      rightState.clearRange(new Instant(0L), ts);
```

Comment:
Don't you want to ts.minus(temporalBound) to be the upper bound of the eviction?

Reply:
OK, bug fixed, testing added. Subclassing `EventTimeEquijoinFn` didn't work because of the DoFn reflection asserts that timer trigger callbacks must be defined int he same class as the class definition. It seems overly restrictive.

Reply:
I think this is basically trying to avoid complexities brought in by subclasses. We might lift such restrictions. You may know that personally I generally dislike implementation inheritance and particularly access to private members of superclasses, so it might just reflect my bias.